app/src/main/java/org/schabi/newpipe/local/feed/service/FeedLoadManager.kt

```diff
@@ -0,0 +1,274 @@
+        feedDatabaseManager.removeOrphansOrOlderStreams()
+
+        FeedEventManager.postEvent(FeedEventManager.Event.SuccessResultEvent(feedResultsHolder.itemsErrors))
+    }.doOnSubscribe {
+        currentProgress.set(-1)
+        maxProgress.set(-1)
+
+        notificationUpdater.onNext(context.getString(R.string.feed_processing_message))
+        FeedEventManager.postEvent(FeedEventManager.Event.ProgressEvent(R.string.feed_processing_message))
+    }.subscribeOn(Schedulers.io())
+
+    private inner class NotificationConsumer : Consumer<Notification<FeedUpdateInfo>> {
+        override fun accept(item: Notification<FeedUpdateInfo>) {
+            currentProgress.incrementAndGet()
+            notificationUpdater.onNext(item.value?.name.orEmpty())
+
+            broadcastProgress()
+        }
+    }
+
+    private inner class DatabaseConsumer : Consumer<List<Notification<FeedUpdateInfo>>> {
+
+        override fun accept(list: List<Notification<FeedUpdateInfo>>) {
+            feedDatabaseManager.database().runInTransaction {
+                for (notification in list) {
+                    when {
+                        notification.isOnNext -> {
+                            val subscriptionId = notification.value.uid
+                            val info = notification.value.listInfo
+
+                            notification.value.newStreamsCount = countNewStreams(info.relatedItems)
+                            feedDatabaseManager.upsertAll(subscriptionId, info.relatedItems)
+                            subscriptionManager.updateFromInfo(subscriptionId, info)
+
+                            if (info.errors.isNotEmpty()) {
+                                feedResultsHolder.addErrors(
+                                    FeedLoadService.RequestException.wrapList(
+                                        subscriptionId,
+                                        info
+                                    )
+                                )
+                                feedDatabaseManager.markAsOutdated(subscriptionId)
+                            }
+                        }
+                        notification.isOnError -> {
+                            val error = notification.error
+                            feedResultsHolder.addError(error)
+
+                            if (error is FeedLoadService.RequestException) {
+                                feedDatabaseManager.markAsOutdated(error.subscriptionId)
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        private fun countNewStreams(list: List<StreamInfoItem>): Int {
+            var count = 0
+            for (item in list) {
+                if (feedDatabaseManager.doesStreamExist(item)) {
+                    return count
+                } else {
+                    count++
+                }
+            }
+            return 0
+        }
```

Comment:
To fix that, we could ensure that the latest stream is somehow stored in the db. I think we have two options here:
- Modify `upsertAll` to keep the latest stream for each subscription.
- Store the UID of the latest stream in the `subscriptions` table.

Reply:
I'd change the `unlinkStreamsOlderThan(offsetTime)` to this by adding the `AND` condition (not tested):
``` SQL
DELETE FROM feed WHERE

        feed.stream_id IN (
            SELECT s.uid FROM streams s

            INNER JOIN feed f
            ON s.uid = f.stream_id

            WHERE s.upload_date < :offsetDateTime
            AND s.uid <> (
                SELECT uid FROM streams
                LEFT JOIN feed ON feed.stream_id = streams.uid
                WHERE feed.subscription_id = f.subscription_id
                ORDER BY upload_date DESC LIMIT 1
            )
                
        )
```

Reply:
Well, why not just considering an item as new if it was uploaded less than 13 weeks ago and it is not in the database? This way we don't need to keep additional data in the db; I don't think it would be a problem not to notify about items that old.

Reply:
Done